textattack/transformations/word_merges/word_merge_masked_lm.py

```diff
@@ -0,0 +1,184 @@
+        """Get replacement words for the word we want to replace using BAE
+        method.
+
+        Args:
+            current_text (AttackedText): Text we want to get replacements for.
+            index (int): index of word we want to replace
+        """
+        masked_texts = []
+        for index in indices_to_modify:
+            temp_text = current_text.replace_word_at_index(
+                index, self._lm_tokenizer.mask_token
+            )
+            masked_texts.append(temp_text.delete_word_at_index(index + 1).text)
+
+        i = 0
+        # 2-D list where for each index to modify we have a list of replacement words
+        replacement_words = []
+        while i < len(masked_texts):
+            inputs = self._encode_text(masked_texts[i : i + self.batch_size])
+            ids = [
+                inputs["input_ids"][i].tolist() for i in range(len(inputs["input_ids"]))
+            ]
+            with torch.no_grad():
+                preds = self._language_model(**inputs)[0]
+
+            for j in range(len(ids)):
+                try:
+                    # Need try-except b/c mask-token located past max_length might be truncated by tokenizer
+                    masked_index = ids[j].index(self._lm_tokenizer.mask_token_id)
+                except ValueError:
+                    replacement_words.append([])
+                    continue
+
+                mask_token_logits = preds[j, masked_index]
+                mask_token_probs = torch.softmax(mask_token_logits, dim=0)
+                ranked_indices = torch.argsort(mask_token_probs)
+                top_words = []
+                for _id in ranked_indices:
+                    _id = _id.item()
+                    token = self._lm_tokenizer.convert_ids_to_tokens(_id)
+                    if utils.is_one_word(token) and not utils.check_if_subword(
+                        token,
+                        self._language_model.config.model_type,
+                        (masked_index == 1),
+                    ):
+                        if mask_token_probs[_id] > self.min_confidence:
+                            top_words.append(token)
+
+                    if len(top_words) >= self.max_candidates:
+                        break
+
+                replacement_words.append(top_words)
+
+            i += self.batch_size
+
+        return replacement_words
+
+    def _get_transformations(self, current_text, indices_to_modify):
+        transformed_texts = []
+        indices_to_modify = list(indices_to_modify)
+        # find indices that are suitable to merge
+        token_tags = [
+            current_text.pos_of_word_index(i) for i in range(current_text.num_words)
+        ]
+        merge_indices = find_merge_index(token_tags)
+        merged_words = self._get_merged_words(current_text, merge_indices)
+        transformed_texts = []
+        for i in range(len(merged_words)):
+            index_to_modify = indices_to_modify[i]
```

Comment:
I think this line should actually be `index_to_modify = merge_indices[i]`. We're trying to merge the indices specified by `merge_indices` and not `indices_to_modify`. Those 2 lists don't necessarily match.

for example, suppose we have the sentence **"I love to play soccer"**, `merge_indices = [3]`, `indices_to_modify = [0,1,2,3,4]`, and `merged_words = ["sleep"]`. Calling `index_to_modify = indices_to_modify[i]` would give us **"sleep to play soccer"**, which is wrong. The correct output should be **"I love to sleep"**.

Reply:
I agree. 

Also, does merge_indices have an order? e.g., which index should be merged first? some indices should have higher priority since they are easily attacked. 

Additionally, could you let me know which line of the code has the function that orders the three perturbations at the same indice? e.g., at [3], maybe the perturbation effect has: Insert > Replace > Merge. In this case, we only use Insert to perturb the text and will discard the other two perturbations.

Reply:
> I agree.
> 
> Also, does merge_indices have an order? e.g., which index should be merged first? some indices should have higher priority since they are easily attacked.
> 
> Additionally, could you let me know which line of the code has the function that orders the three perturbations at the same indice? e.g., at [3], maybe the perturbation effect has: Insert > Replace > Merge. In this case, we only use Insert to perturb the text and will discard the other two perturbations.

`merge_indices` does not have a priority order, since it is just the output of `find_merge_index`. `find_merge_index` runs on the modifiable indices of the text (something like [0,1,2,5]) and is in ascending order of numerical value.

We actually generate all merged sentences first, and then feed them along with all the perturbed sentences generated from `masked_insertion` and `masked_swap` to the search function. So we don't really order the three perturbations and just generate every eligible perturbation, then find the best one using the search function.